store/v2/root/store.go

```diff
@@ -0,0 +1,909 @@
+		}
+
+	case "/subspace":
+		res.Key = req.Data // data holds the subspace prefix
+
+		pairs := kv.Pairs{
+			Pairs: make([]kv.Pair, 0),
+		}
+
+		res.Key = req.Data // data holds the subspace prefix
+
+		iterator := substore.Iterator(res.Key, types.PrefixEndBytes(res.Key))
+		for ; iterator.Valid(); iterator.Next() {
+			pairs.Pairs = append(pairs.Pairs, kv.Pair{Key: iterator.Key(), Value: iterator.Value()})
+		}
+		iterator.Close()
+
+		bz, err := pairs.Marshal()
+		if err != nil {
+			panic(fmt.Errorf("failed to marshal KV pairs: %w", err))
+		}
+
+		res.Value = bz
+
+	default:
+		return sdkerrors.QueryResult(sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "unexpected query path: %v", req.Path), false)
+	}
+
+	return res
+}
+
+func loadSMT(stateCommitmentTxn dbm.DBReadWriter, root []byte) *smt.Store {
+	merkleNodes := prefixdb.NewPrefixReadWriter(stateCommitmentTxn, merkleNodePrefix)
+	merkleValues := prefixdb.NewPrefixReadWriter(stateCommitmentTxn, merkleValuePrefix)
+	return smt.LoadStore(merkleNodes, merkleValues, root)
+}
+
+// Returns closest index and whether it's a match
+func binarySearch(hay []string, ndl string) (int, bool) {
+	var mid int
+	from, to := 0, len(hay)-1
+	for from <= to {
+		mid = (from + to) / 2
+		switch strings.Compare(hay[mid], ndl) {
+		case -1:
+			from = mid + 1
+		case 1:
+			to = mid - 1
+		default:
+			return mid, true
+		}
+	}
+	return from, false
+}
+
+func (pr *prefixRegistry) storeInfo(key string) (sst types.StoreType, ix int, err error) {
+	ix, has := binarySearch(pr.reserved, key)
+	if !has {
+		err = fmt.Errorf("prefix does not exist: %v", key)
+		return
+	}
+	sst, has = pr.StoreSchema[key]
+	if !has {
+		err = fmt.Errorf("prefix is registered but not in schema: %v", key)
+	}
+
+	return
+}
+
+func (pr *prefixRegistry) ReservePrefix(key string, typ types.StoreType) error {
+	if !validSubStoreType(typ) {
+		return fmt.Errorf("StoreType not supported: %v", typ)
+	}
+
+	// Find the neighboring reserved prefix, and check for duplicates and conflicts
+	i, has := binarySearch(pr.reserved, key)
+	if has {
+		return fmt.Errorf("prefix already exists: %v", key)
+	}
+	if i > 0 && strings.HasPrefix(key, pr.reserved[i-1]) {
+		return fmt.Errorf("prefix conflict: '%v' exists, cannot add '%v'", pr.reserved[i-1], key)
+	}
+	if i < len(pr.reserved) && strings.HasPrefix(pr.reserved[i], key) {
+		return fmt.Errorf("prefix conflict: '%v' exists, cannot add '%v'", pr.reserved[i], key)
+	}
+	reserved := pr.reserved[:i]
+	reserved = append(reserved, key)
+	pr.reserved = append(reserved, pr.reserved[i:]...)
+	pr.StoreSchema[key] = typ
+	return nil
+}
+
+func (tlm *traceListenMixin) AddListeners(skey types.StoreKey, listeners []types.WriteListener) {
+	key := skey.Name()
+	if ls, has := tlm.listeners[key]; has {
+		tlm.listeners[key] = append(ls, listeners...)
+	} else {
+		tlm.listeners[key] = listeners
```

Comment:
We can append to `tlm.listeners[key]` whether or not that key already exists in the map (as long as the entire map isn't nil, which it isn't)